### DISCUSSION TITLE
Passing environment variables to LXC

### DISCUSSION THREAD

**Discussion by JasonPyMe** [tteck/Proxmox#2771] [Help Needed]
I am new to LXC, I am more familiar with Docker.  I using the Hompage LXC script it is working fine.

In the Homepage services .conf file I am using variables which works great in Docker but I cannot figure out how to pass variables in LCX.

Any helps would be greatly appreciated.

**Comment by jgonzo-gh** [comment]
I found a [reference](https://linuxcontainers.org/lxc/manpages/man5/lxc.container.conf.5.html#lbBM) to passing variables into an LXC container but init does not seem to pass this to children processes as suggested in the article.  The variable was present in /proc/1/environ but I could not figure out how to parse this and pass it to Homepage.

I did find a different way in which to obfuscate secrets in the Homepage YAML files though.  This [article](https://serverfault.com/questions/413397/how-to-set-environment-variable-in-systemd-service) explains how to set a service variable in systemctl.  I found I could use the command ```systemctl edit homepage``` and create entries for "HOMEPAGE_VAR_XXX" then reference them as {{HOMEPAGE_VAR_XXX}} in the YAML files. I tested with setting a PiHole API variable and the Homepage widget loaded fine.

There is little gained in security because both the YAML files and the systemctl files are world readable thus anybody with access to the container can access your secrets.  The only advantage I see is that you can now share YAML files without exposing secrets.

**Reply by jgault91** [reply]
This, or if you already had a .env or .secrets.env file written, you can follow [this suggestion](https://github.com/tteck/Proxmox/discussions/2549) as well which achieves the same. 

&nbsp; 

I moved from a docker container to a dedicated LXC container using this script, and was able to reuse my existing .secrets.env file to achieve the desired results. 


```bash
touch /opt/homepage/config/.secrets.env
nano /opt/homepage/config/.secrets.env
```

Add in your details

```
HOMEPAGE_VAR_PLEX_URL=your.plexservername.here.com
HOMEPAGE_VAR_NPM_EMAIL=yournpmemail@goeshere.com
HOMEPAGE_VAR_NPM_PWD=yourNPMPasswordHere
HOMEPAGE_VAR_PROXMOX_USERNAME=yourproxmoxusername@pam!proxmox
HOMEPAGE_VAR_PROXMOX_API_KEY=yourProxmoxAPIKey
```

save and quit the file. 

then adjust the systemd service

`nano /etc/systemd/homepage.service`

add your file path line in the `[service]` section

```bash
EnvironmentFile=-/opt/homepage/config/.secrets.env
```

save and quit the file. 

Then in your `services.yaml` you would reference the `{{ HOMEPAGE_VAR_PLEX_URL }}` instead of the url string.

restart the container `reboot` and it should come up. 
 

&nbsp;

And yeah, this is valuable info as well: 

> There is little gained in security because both the YAML files and the systemctl files are world readable thus anybody with access to the container can access your secrets. ***The only advantage I see is that you can now share YAML files without exposing secrets.***

100% agree!
